### DISCUSSION TITLE
What Am I doing wrong here( EntityEncoder Issue)

### DISCUSSION THREAD

**Discussion by SarpongAbasimi** [http4s/http4s#4031] [Q&A]
Hello, please can someone help me to understand what I am doing wrong here. Have some patience with me please, I am just trying to understand what is going on. Thanks 😊

I have this for the route
```    
import org.http4s._
import org.http4s.implicits._
import cats._
import org.http4s.dsl.Http4sDsl
import cats.effect._
import doobie._

def show[F[_]](transactor: Transactor.Aux[F, Unit] )(implicit sync: Sync[F]) = {
    val dsl = new Http4sDsl[F] {}
    import dsl._

    HttpRoutes.of[F]{
      case GET ->  Root/"todos"/IntVar(id) => {
        val todo = TodoHelpers.getTodoByID(id,transactor)
        Ok(todo)
      }
    }
  }
```
and this for Encoding

```
object Encoders {
  final case class Todo(id: Int, todo: String)
  implicit val todoEncoder: Encoder[Todo] = new Encoder[Todo] {
    final def apply(a: Todo): Json = {
      Json.obj(
        ("id", Json.fromInt(a.id)),
        ("todo", Json.fromString(a.todo))
      )
    }
  }
  implicit def todoEntityEncoder[F[_]: Applicative]: EntityEncoder[F, Todo] = jsonEncoderOf[F, Todo]
}
```
But I am getting this error even though I have an `EntityEncoder[F, Todo]`
<img width="779" alt="Screen Shot 2020-12-17 at 22 07 09" src="https://user-images.githubusercontent.com/37377831/102550037-0d388c00-40b5-11eb-9195-898513c10c19.png">

**Comment by hamnis** [comment]
Looks like an import issue.

try adding `import Encoders._`

**Reply by SarpongAbasimi** [reply]
@hamnis so I have already imported Encoders._
which contains these

```
import cats.Applicative
import io.circe._
import org.http4s.circe._
import org.http4s.EntityEncoder

object Encoders {
  final case class Todo(id: Int, todo: String)
  implicit val todoEncoder: Encoder[Todo] = new Encoder[Todo] {
    final def apply(a: Todo): Json = {
      Json.obj(
        ("id", Json.fromInt(a.id)),
        ("todo", Json.fromString(a.todo))
      )
    }
  }
  implicit def todoEntityEncoder[F[_]: Applicative]: EntityEncoder[F, Todo] = jsonEncoderOf[F, Todo]
}
```

Routes contain

```
import org.http4s._
import org.http4s.implicits._
import cats._
import org.http4s.dsl.Http4sDsl
import cats.effect._
import doobie._
import doobie.implicits._
import Encoders._

object Routes {

  def index[F[_]](transactor: Transactor.Aux[F, Unit] )(implicit sync: Sync[F]) = {
    val dsl = new Http4sDsl[F] {}
    import dsl._
    implicit val todoEntityEncoder: EntityEncoder[F, Todo] = ???
    HttpRoutes.of[F]{
      case GET ->  Root => {
        val todos = TodoHelpers.getTodos(transactor)
        Ok("Hey")
      }
    }
  }

  def show[F[_]](transactor: Transactor.Aux[F, Unit] )(implicit sync: Sync[F]) = {
    val dsl = new Http4sDsl[F] {}
    import dsl._

    HttpRoutes.of[F]{
      case GET ->  Root/"todos"/IntVar(id) => {
        val todo = TodoHelpers.getTodoByID(id,transactor)
        Ok(todo)
      }
    }
  }
}
```
And these are my queries

```
import cats.effect._
import doobie._
import doobie.implicits._

final case class Todo(id: Int, todo: String)

object TodoHelpers {
  def getTodos[F[_]](transcator: Transactor.Aux[F, Unit])(implicit b: Bracket[F, Throwable]) = {
    sql"select id, name from todos"
      .query[Todo]
      .stream
      .compile
      .toList
      .transact(transcator)
  }

  def getTodoByID[F[_]](id: Int, transcator: Transactor.Aux[F, Unit])(implicit b: Bracket[F, Throwable]) = {
    sql"select * from todos where id=$id"
      .query[Todo]
      .unique
      .transact(transcator)
  }
}
```
@rossabaker  please I need more explanation  thank you.

**Comment by hamnis** [comment]
if you place your Encoder and the EntityEncoder in the companion object of Todo, they should be found without having to import.

**Comment by rossabaker** [comment]
I'm going to move this to the new discussion board.  Please let us know if that helps, or needs more explanation!

**Reply by SarpongAbasimi** [reply]
@rossabaker please a little bit more explanation as to why this is not working.
Basically the root cause of this issue.

**Comment by Daenyth** [comment]
Try removing the `implicit val todoEntityEncoder: EntityEncoder[F, Todo] = ???` from your Routes.

If you have that *and* the `import Encoders._`, then you have two implicits of the same type, and it becomes ambiguous - the compiler can't pick one, so it fails. I'm a little surprised it didn't say that in the compiler error output

**Reply by SarpongAbasimi** [reply]
Thanks @Daenyth  but that is not the issue.
I put it there just to check if the import from `import Encoders._` was not being picked up.

As you can see the `import Encoders._` is not being picked up 😕

<img width="902" alt="Screen Shot 2020-12-18 at 16 28 04" src="https://user-images.githubusercontent.com/37377831/102637600-23def180-414e-11eb-87f3-b76c77abf40e.png">

**Comment by Daenyth** [comment]
We really can't help you without actual compiler errors. If you right click the missing implicit and use the implicit argument popup, do you see instances there at all?
